finality-provider/service/fp_instance.go

```diff
@@ -530,55 +393,52 @@ func (fp *FinalityProviderInstance) reportCriticalErr(err error) {
 	}
 }
 
-// checkLagging returns true if the lasted voted height is behind by a configured gap
-func (fp *FinalityProviderInstance) checkLagging(currentBlockHeight uint64) bool {
-	return currentBlockHeight >= fp.GetLastProcessedHeight()+fp.cfg.FastSyncGap
-}
-
-// retrySubmitFinalitySignatureUntilBlocksFinalized periodically tries to submit finality signature until success or the block is finalized
+// retrySubmitSigsUntilFinalized periodically tries to submit finality signature until success or the block is finalized
 // error will be returned if maximum retries have been reached or the query to the consumer chain fails
-func (fp *FinalityProviderInstance) retrySubmitFinalitySignatureUntilBlocksFinalized(targetBlocks []*types.BlockInfo) (*types.TxResponse, error) {
+func (fp *FinalityProviderInstance) retrySubmitSigsUntilFinalized(targetBlocks []*types.BlockInfo) (*types.TxResponse, error) {
+	if len(targetBlocks) == 0 {
+		return nil, fmt.Errorf("cannot send signatures for empty blocks")
+	}
+
 	var failedCycles uint32
 	targetHeight := targetBlocks[len(targetBlocks)-1].Height
+
 	// we break the for loop if the block is finalized or the signature is successfully submitted
 	// error will be returned if maximum retries have been reached or the query to the consumer chain fails
 	for {
-		// error will be returned if max retries have been reached
-		var res *types.TxResponse
-		var err error
-		if len(targetBlocks) == 1 {
-			res, err = fp.SubmitFinalitySignature(targetBlocks[0])
-		} else {
+		select {
+		case <-time.After(fp.cfg.SubmissionRetryInterval):
```

Comment:
ok I just found another problem. this will wait `SubmissionRetryInterval` even at the first iteration. Default value is set to 300 seconds. that's why I found out batch sig submission is very slow

I will send out a fix soon

Reply:
https://github.com/babylonlabs-io/finality-provider/pull/217